libsolidity/analysis/TypeChecker.cpp

```diff
@@ -506,26 +471,26 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	else if (_variable.immutable())
 	{
 		if (!_variable.type()->isValueType())
-			m_errorReporter.typeError(6377_error, _variable.location(), "Immutable variables cannot have a non-value type.");
-		if (
-			auto const* functionType = dynamic_cast<FunctionType const*>(_variable.type());
-			functionType && functionType->kind() == FunctionType::Kind::External
-		)
-			m_errorReporter.typeError(3366_error, _variable.location(), "Immutable variables of external function type are not yet supported.");
+			m_errorReporter
+				.typeError(6377_error, _variable.location(), "Immutable variables cannot have a non-value type.");
+		if (auto const* functionType = dynamic_cast<FunctionType const*>(_variable.type());
+			functionType && functionType->kind() == FunctionType::Kind::External)
+			m_errorReporter.typeError(
+				3366_error, _variable.location(), "Immutable variables of external function type are not yet supported."
+			);
```

Comment:
The fact that it's not putting the closing `)` on a new line makes it much harder to see where the condition ends and the body of the function starts. IMO it's a big regression.

Reply:
I actually looked into this a bit last night - and wrote a patch to clang-format adding a ``AlwaysBlockIndent`` mode (https://gist.github.com/ekpyron/07d3796f27527b41784546df20200fda) that'd behave at least *more* like we're used to here - they do specifically exclude control flow constructs from the ``BlockIndent`` mechanism for some reason.

But even with that patch, it still doesn't break after the ``if (`` parentheses (for ``for`` or ``while`` it works), with ``ContinuationIndentWidth: 4`` - with ``ContinuationIndentWidth: 2`` it seems to work (with the patch) - which made me realize: the rationale for our coding style there is rather weak :-).

With a continuation indentation width of 4 spaces (or a 4-space tab), something like
```c++
if (veryLongStatement...)
```
being transformed to
```c++
if (
    veryLongStatement...
)
```
actually doesn't *help* in the first instance, since ``veryLongStatement`` still starts at column 4, so this just adds line breaks without shortening the lines :-). I.e. that's the reason why clang-format is hard to convince to do this - it'll only transform to an improved situation apparently and I haven't found where that's checked (their code is kinda horrible).

Still I actually find
```c++
if (auto const* functionType = dynamic_cast<FunctionType const*>(_variable.type());
        functionType && functionType->kind() == FunctionType::Kind::External
)
```
weirder than
```c++
if (
    auto const* functionType = dynamic_cast<FunctionType const*>(_variable.type());
    functionType && functionType->kind() == FunctionType::Kind::External
)
```

By the way: related:

https://github.com/llvm/llvm-project/issues/79176
https://github.com/llvm/llvm-project/issues/80123
https://github.com/llvm/llvm-project/pull/77699
https://github.com/llvm/llvm-project/issues/67738


Not sure if we should try to pursue this further.

Reply:
Their [tests](https://github.com/llvm/llvm-project/pull/77699/files#diff-3f6f57cda9809a57c5b79e22b4181b3f3aaac7216262d0ef44108f39b0443e9b) are also hideous. If you feel like it'll be worthwhile, go ahead, but I'm honestly quite pessimistic.

Reply:
I mean, even if we did end up with a PR to llvm, that'll still take time to merge (if it's even accepted), to get released, to get packaged... we could use a patched static clang-format in the interim, but yeah... I'd rather take this as evidence that it'd be much easier to try and find a style that's fine with us without this property for now - but you can't say I haven't tried to make this work :-).

Reply:
Then again, the first time we tried this years ago, the current ``BlockIndent`` didn't even exist - https://github.com/llvm/llvm-project/pull/77699 did get merged earlier this year - it's probably not impossible to get this to work. But even if so, I'd rather say we use some other style now and change this part again, once this were to be supported rather than taking this as reason to delay this any further.

Reply:
I think I did get it to work properly, though - wasn't that complicated in the end, the crucial part was just adding
```c++
+  if (Style.AlignAfterOpenBracket == FormatStyle::BAS_AlwaysBlockIndent &&
+      Previous.is(tok::l_paren) &&
+      (getLengthToMatchingParen(Previous, State.Stack) + State.Column - 1 >
+       getColumnLimit(State)))
+    return true;
```
to their "must-break" logic - i.e. just always break at parens if the line gets too long.

And it does get us significantly closer to the current style... ``BreakBinaryOperations: RespectPrecedence`` of unreleased clang 20 also helps a bit and I tweaked a few more options in:
https://github.com/ethereum/solidity/compare/patchedClangFormat

Might be interesting to at least see how llvm/clang reacts if we propose this as a patch - would need some more time to write some tests for it and such, though. They do understandably have rather strict guidelines against this stuff: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options - but we might just qualify with this...